### PR TITLE
std::core: rewrite DString

### PR DESCRIPTION
**Pull request by pierrec**
This PR aims at making the use of DString easier. More precisely, now a DString can be created and built and the resulting string from str_view() or zstr_view() can be freed without having to keep a reference to the original DString.

### CODE REVIEW

File: lib/std/core/dstring.c3 [hunk 1]
```diff
@@ -10,37 +20,43 @@ const usz MIN_CAPACITY @private = 16;
 fn void DString.init(&self, usz capacity = MIN_CAPACITY, Allocator* using = mem::heap())
 {
 	if (capacity < MIN_CAPACITY) capacity = MIN_CAPACITY;
-	StringData* data = malloc(StringData, 1, .using = using, .end_padding = capacity);
-	data.allocator = using;
-	data.len = 0;
-	data.capacity = capacity;
-	*self = (DString)data;
+	char* start = malloc(char, capacity + StringData.sizeof, .using = using);
```
Comment (lerno) [inline lib/std/core/dstring.c3:23]: This doesn't guarantee alignment.

**Comment by lerno** [comment]
Let me look at rewriting it, as there are some gnarly things to consider. Keep the conv.c3 update though. If you have time, look at converting Streamable to a protocol.

**Comment by lerno** [comment]
Uh, I also realize that it won't work.

**Comment by pierrec** [comment]
> Let me look at rewriting it, as there are some gnarly things to consider. Keep the conv.c3 update though. If you have time, look at converting Streamable to a protocol.

Will have a look.

Thanks for considering this approach!

**Pull request closed without merging**